README.md
# 简介

基于 Autojs 的蚂蚁森林自动收能量脚本，采用 4.1.1 Alpha 版本开发。解锁模块参考自：https://github.com/e1399579/autojs

交流群：524611323

# 使用

下载安装 [Autojs](https://github.com/hyb1996/Auto.js) 之后把整个脚本项目放进 __"/sdcard/脚本/"__ 文件夹下面。运行项目或者 main 即可。

# 功能

- 自动匹配不同系统下自动化的方式，安卓7及以上通过无障碍服务模拟操作，以下版本通过 root 权限模拟操作；
- 自动识别屏幕锁定方式并根据配置的密码解锁，支持图形解锁，PIN解锁，混合密码解锁；
- 识别自己能量球的倒计时，和好友列表中的倒计时做对比，取最小值作为下次收取的等待时间；
- 识别好友能量罩，下一次收取时跳过开启能量罩的好友；
- 默认使用倒计时收取，可通过配置打开循环收取；
- 根据设置选择是否帮助好友收取能量；
- 根据白名单实现不收取特定好友能量；
- 收取完毕后悬浮框显示收取的能量数量；
- 可以直接更新最新脚本。
- 收取过程中，添加统计悬浮信息
- 倒计时模式下，自动添加定时任务，如果崩溃可以自动启动
- UI组件文字可配置，在ui_config中修改

# 配置

运行 config.js 后可以看到如下图所示的配置：

![配置图片](./config.jpg)

- 执行模式：
  - 计时：由程序自动计算下次收取所需要的时间，通过设置最大等待时间来限制执行。
  - 循环：循环执行，选择循环模式时可以设置循环次数。
- 帮好友收取：是否帮好友收取已经成熟的能量。
- 颜色偏移量：如果识别失败可以尝试增加该值，默认为50，即80%的相似度。
- 解锁密码：手机解锁密码，如果是图形解锁则为图形经过的点对应的数字。
- 解锁操作时延：解锁模块的延时，解锁操作过快导致出错时可修改，默认为1000ms。
- 控件搜索超时：控件搜索时最大搜索时间，找不到控件时可修改，默认为1000ms。
- 白名单：将好友的昵称添加到白名单实现不收取特定好友的能量。
- 清除本地储存：清除储存在本地的配置，相当于初始化。
## 新增配置信息, 基本一次设置或者说默认设置就能运行的
```javascript
vsr default_conf = {
  ...
  // 是否显示debug详细日志
  show_debug_info: true,
  // 预加载排行榜数据的超时时间 正常情况下100个好友约2000ms，超时时间配置为实际加载时间的2-3倍，具体自己计算
  timeoutLoadFriendList: 6000,
  // 是否永不停止，仅计时模式有效
  never_stop: false,
  // 永不停止重新激活时间。当获取倒计时低于该值时睡眠reactive_time分钟
  reactive_time: 30
};
// 不通过可视化配置的值，即需要直接修改config.js中的信息，而不是运行config.js之后配置
var non_gui_config = {
  // 等待列表稳定的计数，越大越慢但是越稳定，越小越快但是容易导致漏收
  friendListStableCount: 3,
  // 滑动开始距离底部的高度
  bottomHeight: 150,
  // 最大重试次数
  maxRetryTime: 5,
  // mini悬浮窗的位置
  min_floaty_x: 145,
  min_floaty_y: 20,
  // 帮助收取能量球颜色, 可以多增加几组以便全部能够找到，但是越多识别速度越慢
  help_energy_ball_color: ['#f99236', '#f7af70'],
  // 保存日志文件 文件名 log-verboses.log
  save_log_file: false,
  // 列表下滑执行速度 毫秒
  scroll_down_speed: 200
}

var ui_config = {
  // 是否进入个人页面用
  home_ui_content: '背包|通知',
  // 是否进入好友页面用
  friend_home_ui_content: '浇水|发消息',
  // 是否进入排行榜用
  friend_list_ui_content: '好友排行榜',
  // 校验排行榜加载完毕用
  no_more_ui_content: '没有更多了',
  // 检测是否存在可收取能量球
  collectable_energy_ball_content: /.*克/
}
```


# 添加解锁设备

在 Unlock.js 中，按照以下格式扩展：

```javascript
var Devices = { 
  device_1: function(obj) {
    this.__proto__ = obj;

    this.unlock = function(password) {
      if (typeof password !== "string") throw new Error("密码应为字符串！");
      
      // 此处为解锁的代码

      return this.check_unlock();
    }
  },
  device_2: function(obj) {
    ...
  },
  device_3: function(obj) {
    ...
  }
}
```

上述所示为最简单的解锁模板，也可以参考 Unlocl.js 默认多解锁方式的代码进行修改。

然后在下方的 MyDevice 中设置解锁设备：

```javascript
var MyDevice = Devices.device_1;
```

# 注意事项

解锁仅支持：

- 具有ROOT权限的安卓5.0及以上版本
- 没有ROOT权限的安卓7.0及以上版本

# 目前存在的问题

- Autojs 在锁屏状态下由于软件优先度被降低导致 sleep() 函数时间不准确
- 当好友列表数量超过大约 100 后，列表下拉会很慢

# CHANGELOG

- 2019/1/31 
  - ~~发现识别能量罩时采用的函数不合适~~（使用了同步获取 toast 的方法，会卡住，已修正）
  - ~~帮助好友收取时，默认所有能量球都各点击一遍，效率太低~~（已修正）
  - 重构代码，添加注释

- 2019/2/1
  - ~~Toast 监听器超过10过导致报错~~（已修正）
  - ~~帮助好友收取时有时候会失败~~（因为控件下方文字闪烁导致，已修正）
  - 不限制监听器数量并且每次运行完成后清空监听器
  
- 2019/2/2
  - ~~自己的倒计时减为0时会结束收取而不是立马收取下一次~~（已修正）
  
- 2019/2/5
  - ~~实际运行中安卓7.0以下会报错~~（已修正）

- 2019/2/24
  - 多语言问题，繁体或者英文环境下判断字符不同（取消）
  - ~~当收取次数设置为 0 次时，收取行为出错~~（已修正）
  - ~~初次进入蚂蚁森林弹窗提醒添加至首页和合种信息~~（已修正）

- 2019/3/5
  - 重构了一下 Unlock 的内容，方便添加设备
  - 由于基本所有设备解锁都有滑动层，因此去掉了判断是否有滑动层的代码
  - 目前看来新版本效果不错，因此去掉了 old 版本的脚本
  - 增加循环收取的功能

- 2019/3/7
  - 增加白名单功能
  - 计算颜色相似度，修改默认颜色偏移量为50

- 2019/3/12
  - 增加设置UI界面
  - 解决春种活动导致的BUG

- 2019/3/13
  - 增加脚本更新功能

- 2019/3/14
  - 修复好友列表因网络加载慢出现“正在加载”时报错的问题
  - 修复循环执行时的各种问题

- 2019/3/16
  - 实现真正的脚本更新功能

- 2019/3/20
  - 重复运行脚本会报错，因此增加脚本重复运行检查功能

- 2019/7/4
  - 计时模式下自动创建定时任务
  - UI查找文字可配置
  - 失败重试机制，脚本运行更加健壮
  - 其他优化

CHANGELOG.md
# 2.1.4
- 计时模式下自动创建定时任务
- UI查找文字可配置
- 失败重试机制，脚本运行更加健壮
- 其他优化

config.js
/*
 * @Author: NickHopps
 * @Last Modified by: TonyJiangWJ
 * @Last Modified time: 2019-07-04 00:14:13
 * @Description: 配置文件
 */

"ui";
let {
  debugInfo, logInfo, infoLog, warnInfo, errorInfo
} = require('./lib/LogUtils.js')

var default_conf = {
  color_offset: 50,
  password: "",
  help_friend: true,
  is_cycle: false,
  cycle_times: 10,
  delay_unlock: 1000,
  timeout_findOne: 1000,
  max_collect_wait_time: 20,
  white_list: [],
  // 是否显示debug详细日志
  show_debug_info: true,
  // 预加载排行榜数据的超时时间 正常情况下100个好友约2000ms，超时时间配置为实际加载时间的2-3倍，具体自己计算
  timeoutLoadFriendList: 6000,
  // 是否永不停止，仅计时模式有效
  never_stop: false,
  // 永不停止重新激活时间。当获取倒计时低于该值时睡眠reactive_time分钟
  reactive_time: 30
};

var non_gui_config = {
  // 等待列表稳定的计数，越大越慢但是越稳定，越小越快但是容易导致漏收
  friendListStableCount: 3,
  // 滑动开始距离底部的高度
  bottomHeight: 150,
  // 最大重试次数
  maxRetryTime: 5,
  // mini悬浮窗的位置
  min_floaty_x: 145,
  min_floaty_y: 20,
  // 帮助收取能量球颜色, 可以多增加几组以便全部能够找到，但是越多识别速度越慢
  help_energy_ball_color: ['#f99236', '#f7af70'],
  // 保存日志文件 文件名 log-verboses.log
  save_log_file: true,
  // 列表下滑执行速度 毫秒
  scroll_down_speed: 100
}

// 使用字符串方式的正则表达式 不要使用/pattern/方式 否则无法生效
var ui_config = {
  // 是否进入个人页面用
  home_ui_content: '背包|通知',
  // 是否进入好友页面用
  friend_home_ui_content: '浇水|发消息',
  // 是否进入排行榜用
  friend_list_ui_content: '好友排行榜',
  // 校验排行榜加载完毕用
  no_more_ui_content: '没有更多了',
  // 检测是否存在可收取能量球
  collectable_energy_ball_content: '.*\\d+克'
}

var config = storages.create("ant_forest_config");
if (!config.contains("color_offset")) {
  logInfo("使用默认配置", true);
  // 储存默认配置到本地
  Object.keys(default_conf).forEach(function (key) {
    config.put(key, default_conf[key]);
  });
} else {
  // 某些非可视化配置的
  Object.keys(default_conf).forEach(key => {
    let storedConfigItem = config.get(key)
    if (storedConfigItem === undefined) {
      storedConfigItem = default_conf[key]
    }
    config.put(key, storedConfigItem)
  })
}
// UI配置直接设置到storages
Object.keys(ui_config).forEach(key => {
  config.put(key, ui_config[key])
})

// 非GUI可配置的配置直接设置到storages
Object.keys(non_gui_config).forEach(key => {
  config.put(key, non_gui_config[key])
})

function draw_view() {
  ui.layout(
    <ScrollView>
      <vertical id="viewport">
        <frame>
          <appbar>
            <toolbar id="toolbar" title="执行配置" />
          </appbar>
        </frame>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="设置执行模式：" textColor="#666666" textSize="14sp" />
          <radiogroup id="exec_pattern" orientation="horizontal" margin="0 10">
            <radio text="计时" checked="{{!config.get('is_cycle')}}" />
            <radio text="循环" checked="{{config.get('is_cycle')}}" marginLeft="20" />
          </radiogroup>
          <vertical visibility="{{config.get('is_cycle') ? 'visible' : 'gone'}}" w="*" gravity="left" layout_gravity="left">
            <text text="循环次数：" textColor="#666666" textSize="14sp" />
            <input id="cycle_times" inputType="number" text="{{config.get('cycle_times')}}" />
          </vertical>
          <vertical  w="*" visibility="{{config.get('is_cycle') ? 'gone' : 'visible'}}" >
            <text text="永不停止：" textColor="#666666" textSize="14sp" />
            <radiogroup id="never_stop" orientation="horizontal" margin="0 10">
              <radio text="是" checked="{{config.get('never_stop')}}" />
              <radio text="否" checked="{{!config.get('never_stop')}}" marginLeft="20" />
            </radiogroup>
          </vertical>
          <vertical visibility="{{(!config.get('is_cycle')) ? 'visible': 'gone'}}" w="*">
            <vertical visibility="{{config.get('never_stop') ? 'visible' : 'gone'}}" w="*" gravity="left" layout_gravity="left">
              <text text="重新激活等待时间：" textColor="#666666" textSize="14sp" />
              <input id="reactive_time" inputType="number" text="{{config.get('reactive_time')}}" />
            </vertical>
          </vertical>
        </vertical>
        <horizontal w="*" h="1sp" bg="#cccccc" margin="10 0"></horizontal>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="帮好友收取：" textColor="#666666" textSize="14sp" />
          <radiogroup id="is_help_fris" orientation="horizontal" margin="0 10">
            <radio text="是" checked="{{config.get('help_friend')}}" />
            <radio text="否" checked="{{!config.get('help_friend')}}" marginLeft="20" />
          </radiogroup>
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="显示debug日志：" textColor="#666666" textSize="14sp" />
          <radiogroup id="show_debug_info" orientation="horizontal" margin="0 10">
            <radio text="是" checked="{{config.get('show_debug_info')}}" />
            <radio text="否" checked="{{!config.get('show_debug_info')}}" marginLeft="20" />
          </radiogroup>
        </vertical>
        <horizontal w="*" h="1sp" bg="#cccccc" margin="10 0"></horizontal>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="颜色偏移量：" textColor="#666666" textSize="14sp" />
          <input id="color_offset" inputType="number" text="{{config.get('color_offset')}}" />
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="解锁密码：" textColor="#666666" textSize="14sp" />
          <input id="password" inputType="textPassword" text="{{config.get('password')}}" />
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="最大等待时间（分钟）：" textColor="#666666" textSize="14sp" />
          <input id="max_collect_wait_time" inputType="number" text="{{config.get('max_collect_wait_time')}}" />
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="好友列表预加载时延：" textColor="#666666" textSize="14sp" />
          <input id="timeoutLoadFriendList" inputType="number" text="{{config.get('timeoutLoadFriendList')}}" />
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="解锁操作时延：" textColor="#666666" textSize="14sp" />
          <input id="delay_unlock" inputType="number" text="{{config.get('delay_unlock')}}" />
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="控件搜索超时：" textColor="#666666" textSize="14sp" />
          <input id="timeout_findOne" inputType="number" text="{{config.get('timeout_findOne')}}" />
        </vertical>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <text text="白名单：" textColor="#666666" textSize="14sp" />
          <text visibility="{{config.get('white_list').length == 0 ? 'visible' : 'gone'}}" w="*" h="80" gravity="center" layout_gravity="center" text="白名单为空" textColor="#999999" textSize="18sp" margin="0 20" bg="#eeeeee" />
          <frame>
            <list id="white_list">
              <horizontal w="*" h="40" gravity="left" bg="#efefef" margin="0 5">
                <text id="name" layout_weight='1' h="30" gravity="left|center" layout_gravity="left|center" textSize="16sp" text="{{name}}" margin="10 0" />
                <card id="delete" w="30" h="30" cardBackgroundColor="#fafafa" cardCornerRadius="15dp" layout_gravity="center" marginRight="10">
                  <text textSize="16dp" textColor="#555555" gravity="center">×</text>
                </card>
              </horizontal>
            </list>
          </frame>
          <button w="*" id="add" text="添加" gravity="center" layout_gravity="center" />
        </vertical>
        <horizontal w="*" h="1sp" bg="#cccccc" margin="10 0"></horizontal>
        <vertical w="*" gravity="left" layout_gravity="left" margin="10">
          <button w="*" id="clear" text="清除本地储存" gravity="center" layout_gravity="center" />
        </vertical>
      </vertical>
    </ScrollView>
  );

  // 更新本地配置同时重绘UI
  function update(target, new_val) {
    config.put(target, new_val);
    if (target == "is_cycle" || target == "white_list" || target == 'never_stop') draw_view();
  }

  // 格式化
  function format(val) {
    return val.toString();
  }

  // 更新选中的执行方法
  ui.exec_pattern.setOnCheckedChangeListener(function (radioGroup, id) {
    let index = (id + 1) % radioGroup.getChildCount();
    //toast(radioGroup.getChildAt(index).getText());
    if (radioGroup.getChildAt(index).getText() == "循环") {
      update("is_cycle", true);
    } else {
      update("is_cycle", false);
    }
  });

  // 更新是否帮助好友
  ui.is_help_fris.setOnCheckedChangeListener(function (radioGroup, id) {
    let index = (id + 1) % radioGroup.getChildCount();
    //toast(radioGroup.getChildAt(index).getText());
    if (radioGroup.getChildAt(index).getText() == "是") {
      update("help_friend", true);
    } else {
      update("help_friend", false);
    }
  });

  // 更新是否显示debug日志
  ui.show_debug_info.setOnCheckedChangeListener(function (radioGroup, id) {
    let index = (id + 1) % radioGroup.getChildCount();
    //toast(radioGroup.getChildAt(index).getText());
    if (radioGroup.getChildAt(index).getText() == "是") {
      update("show_debug_info", true);
    } else {
      update("show_debug_info", false);
    }
  });

  ui.never_stop.setOnCheckedChangeListener(function (radioGroup, id) {
    let index = (id + 1) % radioGroup.getChildCount();
    //toast(radioGroup.getChildAt(index).getText());
    if (radioGroup.getChildAt(index).getText() == "是") {
      update("never_stop", true);
    } else {
      update("never_stop", false);
    }
  });

  // 更新颜色偏移
  ui.emitter.on("pause", () => {
    if (config.contains("color_offset")) {
      update("cycle_times", format(ui.cycle_times.getText()));
      update("color_offset", format(ui.color_offset.getText()));
      update("password", format(ui.password.getText()));
      update("max_collect_wait_time", format(ui.max_collect_wait_time.getText()));
      update("delay_unlock", format(ui.delay_unlock.getText()));
      update("timeout_findOne", format(ui.timeout_findOne.getText()));
      update("timeoutLoadFriendList", format(ui.timeoutLoadFriendList.getText()));
      update("reactive_time", format(ui.reactive_time.getText()));
    }
  });

  // 白名单缓存
  var list_temp = config.get("white_list").map(i => { return { name: i } });
  // 生成白名单
  ui.white_list.setDataSource(list_temp);
  // 从白名单中删除
  ui.white_list.on("item_bind", function (itemView, itemHolder) {
    itemView.delete.on("click", function () {
      list_temp.splice(itemHolder.position, 1);
      update("white_list", list_temp.map(i => i['name']));
    });
  });
  // 添加到白名单
  ui.add.on("click", () => {
    dialogs.rawInput("请输入好友昵称")
      .then(fri_name => {
        if (!fri_name) return;
        list_temp.push({ name: fri_name });
        update("white_list", list_temp.map(i => i['name']));
      });
  });

  // 清除本地储存
  ui.clear.on("click", () => {
    confirm("确定要清除本地储存吗？")
      .then(ok => {
        if (ok) {
          storages.remove("ant_forest_config");
          toastLog("清除成功");
        }
      });
  });
}

draw_view();